src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,435 @@
+                '_id' => (string) $key,
+                'token' => $token,
+                'expires_at' => $this->createDateTime($now + $this->initialTtl),
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $key->reduceLifetime($this->initialTtl);
+
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (WriteException $e) {
+            if ($this->isDuplicateKeyException($e)) {
+                throw new LockConflictedException('Lock was acquired by someone else', 0, $e);
+            }
+            throw new LockAcquiringException('Failed to acquire lock', 0, $e);
+        }
+
+        if ($key->isExpired()) {
+            throw new LockExpiredException(sprintf('Failed to store the "%s" lock.', $key));
+        }
+
+        if ($this->options['createTtlIndexProbability'] > 0.0
+            && (
+                1.0 === $this->options['createTtlIndexProbability']
+                || (random_int(0, PHP_INT_MAX) / PHP_INT_MAX) <= $this->options['createTtlIndexProbability']
+            )
+        ) {
+            if ($this->getDatabaseVersion() < '2.2') {
+                $this->removeExpiredLocks();
+            } else {
+                $this->createTtlIndex();
+            }
+        }
+    }
+
+    /**
+     * Remove expired locks from the collection. This is a fallback for MongoDB
+     * servers prior to version 2.2. For versions 2.2+ self::createTtlIndex is
+     * called instead based on the createTtlIndexProbability options.
+     */
+    private function removeExpiredLocks()
+    {
+        $now = microtime(true);
+
+        $filter = array(
+            'expires_at' => array(
+                '$lt' => $this->createDateTime($now),
+            ),
+        );
+
+        $this->getCollection()->deleteMany($filter);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function waitAndSave(Key $key)
+    {
+        throw new NotSupportedException(sprintf(
+            'The store "%s" does not support blocking locks.',
+            __CLASS__
+        ));
+    }
+
+    /**
+     * @return bool
+     */
+    private function isDuplicateKeyException(
+        WriteException $e
+    ): bool {
+        $writeErrors = $e->getWriteResult()->getWriteErrors();
+        if (1 === \count($writeErrors)) {
+            $code = $writeErrors[0]->getCode();
+        } else {
+            $code = $e->getCode();
+        }
+        // Mongo error E11000 - DuplicateKey
+        return 11000 === $code;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockStorageException
+     * @throws LockExpiredException
+     */
+    public function putOffExpiration(Key $key, $ttl)
+    {
+        $now = microtime(true);
+
+        $filter = array(
+            '_id' => (string) $key,
+            'token' => $this->getUniqueToken($key),
+            'expires_at' => array(
+                '$gte' => $this->createDateTime($now),
```

Comment:
should be `token = $token OR expires_at <= now()` Look at PdoStore for consistency

Reply:
I have consolidated `save` and `putOffExpiration` into a common method since they are now the same DB statement